### PR TITLE
phpunit 10 support proposal

### PR DESCRIPTION
**Pull request by raneomik**
proposal for #118 issue

### CODE REVIEW

**Comment by Chris53897** [comment]
thanks for your work @raneomik 
Could you change the [PHPUnit](https://phpunit.de/) version in composer.json to  "phpunit/phpunit": "^9.5|^10.4",  ?
I guess this would trigger the ci to test with PHPUnit 10 if possible.

Or did i oversee a testrun fo PHPUnit 10?

**Comment by raneomik** [comment]
> thanks for your work @raneomik Could you change the [PHPUnit](https://phpunit.de/) version in composer.json to "phpunit/phpunit": "^9.5|^10.4", ? I guess this would trigger the ci to test with PHPUnit 10 if possible.
> 
> Or did i oversee a testrun fo PHPUnit 10?

besides  some deprecations and a different phpunit config needed /suggested (does CI support a dedicated "phpunit10.xml" file ?), all seems good

**Comment by kbond** [comment]
This looks great @raneomik! Is enabling the extension in phpunit 10 identical to: https://github.com/zenstruck/browser#installation?

**Comment by raneomik** [comment]
@kbond yes, with a small adaptation  required by phpunit 10 ("bootstrap" instead of "extension"), like this :

```xml
<phpunit>
    ...
    <extensions>
        <bootstrap class="Zenstruck\Browser\Test\BrowserExtension" />
    </extensions>
</phpunit>
````

might be worth mentioning it in the readme, like it is the case for example in this extension  : 
https://github.com/dmaicher/doctrine-test-bundle#using-the-bundle-with-phpunit ?

File: src/Browser/Test/LegacyExtension.php [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                // noop - swallow exceptions related to dumping the current state so as to not
+                // lose the actual error/failure.
+            }
+        }
+    }
+
+    private static function normalizeTestName(string $name): string
+    {
+        // Try to match for a numeric data set index. If it didn't, match for a string one.
+        if (!\preg_match('#^([\w:\\\]+)(.+\#(\d+))#', $name, $matches)) {
```
Comment (raneomik) [inline src/Browser/Test/LegacyExtension.php:117]: it needed a slight adapt here, it was `#^([\w:\\\]+)(.+\#(\d+).+)?$#` & `#^([\w:\\\]+)(.+"([\w ]+)".+)?$#` to extract from names like `Namespace\TestClass::testMethod with data set #0 ('test data')` / `data set "set name" (test data)`.
Under phpunit 10 it returns names without "(test data)" (like `Namespace\TestClass::testMethod with data set #0` / `data set "set name"`)

Comment (raneomik) [reply src/Browser/Test/LegacyExtension.php:117]: A test should be written to check if both versions (above and below phpunit 10 ) work fine?

Comment (raneomik) [reply src/Browser/Test/LegacyExtension.php:117]: @nikophil indeed, I tested it a bit and there's a side effect : the case without datasets (`Namespace\TestClass::testMethod` for example) that isn't captured by this...
I put the PR as draft before finding the solution ...

Are those the possible cases to capture or are there more ? : 

- `Namespace\TestClass::testMethod`
- `Namespace\TestClass::testMethod with data set #0 (data)`
- `Namespace\TestClass::testMethod with data set "test" (data)`

new in phpunit 10 (checked but didn't see if `PHPUnit\Event\Code\Test` provides current dataset, only its name or the whole set in the dataset method ...) : 
- `Namespace\TestClass::testMethod with data set #0`
- `Namespace\TestClass::testMethod with data set "test"`

Comment (raneomik) [reply src/Browser/Test/LegacyExtension.php:117]: added a missing "``?``", which should capture these 5 cases

File: phpunit.xml.dist.10
Comment (nikophil) [inline phpunit.xml.dist.10:1]: I would have call this file something like `phpunit-10.xml.dist` just because in my IDE I have associated `*.xml.dist` to "xml", and then we get color highlighting out of the box

**Review by nikophil — COMMENTED** [review]
hi @raneomik 

thanks for this work!

sadly, I've still not used phpunit 10, so I don't understand fully this PR.

I'm wondering if we should not change the CI to test with both phpunit version?

WDYT?

**Comment by raneomik** [comment]
Hello @nikophil,

> sadly, I've still not used phpunit 10, so I don't understand fully this PR.

I really not am fluent in using it either, it's the first project with functional tests using zenstruck/browser which we are trying to migrate to phpunit 10... this PR basicaly just takes what was done in other phpunit extension projects to make it compatible with phpunit10 & previous version at the same time (for example : [panther](https://github.com/symfony/panther/blob/main/src/ServerExtension.php), [DAMA](https://github.com/dmaicher/doctrine-test-bundle/blob/master/src/DAMA/DoctrineTestBundle/PHPUnit/PHPUnitExtension.php), [speedtrap example](https://localheinz.com/articles/2023/02/14/extending-phpunit-with-its-new-event-system/#content-implementing-an-extension-in-the-new-event-system) referenced in #118...) (including some other updates to drop some phpunit10 deprecations in its own tests - aka static dataproviders) 

> I'm wondering if we should not change the CI to test with both phpunit version ?

it is exactly what I was wondering [here](https://github.com/zenstruck/browser/pull/137#issuecomment-1834081635) 😅 
The update in composer.json suggested by @Chris53897 already makes the CI use phpunit 10, only with some deprecations and it doesn't seem to affect the coverage reports. Adding this file will drop [this deprecation/new file suggest](https://github.com/zenstruck/browser/actions/runs/7084716495/job/19279534489?pr=137#step:2:551) but only in the job that use the phpunit10 dedicated file ([panther's phpunit 10 in CI](https://github.com/symfony/panther/blob/main/.github/workflows/ci.yml#L226) for example)

File: phpunit-10.xml.dist [hunk 2]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kbond) [inline phpunit-10.xml.dist]: Since this file is not being used by default and the tests seem to pass (just config file deprecations), should we remove?

Alternatively, I think if we rename this to phpunit.dist.xml it will be picked up by phpunit 10-only (need to confirm).

**Comment by kbond** [comment]
> might be worth mentioning it in the readme

Do you mind adding in this PR?

File: phpunit-10.xml.dist (hunk 2 above)
Comment (raneomik) [reply phpunit-10.xml.dist]: Actually, under phpunit 9 it triggers a warning but does execute tests with no problems. (cf.: [CI](https://github.com/zenstruck/browser/actions/runs/7171192812/job/19525600507?pr=137#step:2:640))

File: src/Browser/Test/LegacyExtension.php (hunk 1 above)
Comment (raneomik) [reply src/Browser/Test/LegacyExtension.php:117]: since the support for phpunit 10 test outputs, without drops on previous version's formats, seems ok here (not totaly tested the edge cases mentioned in #67 but doesn't seem the topic here), can I resolve this ?

**Review by kbond — APPROVED** [review]

**Comment by kbond** [comment]
Thanks for this @raneomik!
